File: Backend/app/models/ticket_strategy.py
"""
Ticket Strategy: a reusable ticketing template owned by an organizer.
Each strategy has named tiers (e.g. Platinum, Diamond) with prices.
Strategies are linked to events — when an event uses a strategy,
TicketTier rows are copied from the strategy tiers on event creation.
"""
from datetime import datetime
from sqlalchemy import String, Text, Integer, DateTime, ForeignKey
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base


class TicketStrategy(Base):
    """Reusable ticketing template created by organizers (like venues)."""
    __tablename__ = "ticket_strategies"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    organizer_id: Mapped[int] = mapped_column(ForeignKey("users.id"), nullable=False, index=True)
    name: Mapped[str] = mapped_column(String(255), nullable=False)  # e.g. "Concert Standard", "Gala VIP"
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=datetime.utcnow)
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=datetime.utcnow, onupdate=datetime.utcnow)

    organizer = relationship("User", back_populates="ticket_strategies")
    tiers = relationship("TicketStrategyTier", back_populates="strategy", cascade="all, delete-orphan",
                         order_by="TicketStrategyTier.display_order")
    events = relationship("Event", back_populates="ticket_strategy")


class TicketStrategyTier(Base):
    """A tier within a ticket strategy template."""
    __tablename__ = "ticket_strategy_tiers"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    strategy_id: Mapped[int] = mapped_column(ForeignKey("ticket_strategies.id"), nullable=False, index=True)
    name: Mapped[str] = mapped_column(String(64), nullable=False)  # e.g. "Platinum", "Diamond", "General"
    description: Mapped[str | None] = mapped_column(Text, nullable=True)  # what this tier provides
    price_cents: Mapped[int] = mapped_column(Integer, nullable=False)
    display_order: Mapped[int] = mapped_column(Integer, nullable=False, default=0)

    strategy = relationship("TicketStrategy", back_populates="tiers")

File: FrontEnd/lib/models/ticket_strategy.dart
class TicketStrategyTier {
  final int id;
  final String name;
  final String? description;
  final int priceCents;
  final int displayOrder;

  TicketStrategyTier({
    required this.id,
    required this.name,
    this.description,
    required this.priceCents,
    this.displayOrder = 0,
  });

  factory TicketStrategyTier.fromJson(Map<String, dynamic> json) {
    return TicketStrategyTier(
      id: json['id'],
      name: json['name'],
      description: json['description'],
      priceCents: json['price_cents'],
      displayOrder: json['display_order'] ?? 0,
    );
  }

  bool get isFree => priceCents == 0;
  String get priceFormatted =>
      isFree ? 'FREE' : '\$${(priceCents / 100).toStringAsFixed(2)}';
}


class TicketStrategy {
  final int id;
  final int organizerId;
  final String name;
  final List<TicketStrategyTier> tiers;
  final DateTime createdAt;
  final DateTime updatedAt;

  TicketStrategy({
    required this.id,
    required this.organizerId,
    required this.name,
    required this.tiers,
    required this.createdAt,
    required this.updatedAt,
  });

  factory TicketStrategy.fromJson(Map<String, dynamic> json) {
    return TicketStrategy(
      id: json['id'],
      organizerId: json['organizer_id'],
      name: json['name'],
      tiers: (json['tiers'] as List? ?? [])
          .map((t) => TicketStrategyTier.fromJson(t))
          .toList()
        ..sort((a, b) => a.displayOrder.compareTo(b.displayOrder)),
      createdAt: DateTime.parse(json['created_at']),
      updatedAt: DateTime.parse(json['updated_at']),
    );
  }

  /// Summary line: "3 tiers: General ($10), VIP ($50), ..."
  String get tiersSummary {
    if (tiers.isEmpty) return 'No tiers';
    final items = tiers.map((t) => '${t.name} (${t.priceFormatted})').join(', ');
    return '${tiers.length} tier${tiers.length == 1 ? '' : 's'}: $items';
  }
}
